controllers/machinedeployment_sync_test.go

```diff
@@ -0,0 +1,223 @@
+					Generation: 2,
+				},
+				Spec: clusterv1.MachineDeploymentSpec{
+					Replicas: &replicas,
+				},
+			},
+			expectedStatus: clusterv1.MachineDeploymentStatus{
+				ObservedGeneration:  2,
+				Replicas:            2,
+				UpdatedReplicas:     2,
+				ReadyReplicas:       1,
+				AvailableReplicas:   1,
+				UnavailableReplicas: 1,
+				Phase:               "scaling up",
+			},
+		},
+		{
+			desc: "scaling down",
+			machineSets: []*clusterv1.MachineSet{{
+				Spec: clusterv1.MachineSetSpec{
+					Replicas: &replicas,
+				},
+				Status: clusterv1.MachineSetStatus{
+					Selector:           "",
+					AvailableReplicas:  3,
+					ReadyReplicas:      2,
+					Replicas:           2,
+					ObservedGeneration: 1,
+				},
+			}},
+			newMachineSet: &clusterv1.MachineSet{
+				Spec: clusterv1.MachineSetSpec{
+					Replicas: &replicas,
+				},
+				Status: clusterv1.MachineSetStatus{
+					Selector:           "",
+					AvailableReplicas:  3,
+					ReadyReplicas:      2,
+					Replicas:           2,
+					ObservedGeneration: 1,
+				},
+			},
+			deployment: &clusterv1.MachineDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Generation: 2,
+				},
+				Spec: clusterv1.MachineDeploymentSpec{
+					Replicas: &replicas,
+				},
+			},
+			expectedStatus: clusterv1.MachineDeploymentStatus{
+				ObservedGeneration:  2,
+				Replicas:            2,
+				UpdatedReplicas:     2,
+				ReadyReplicas:       2,
+				AvailableReplicas:   3,
+				UnavailableReplicas: 0,
+				Phase:               "scaling down",
+			},
+		},
+		{
+			desc: "machine set failed",
+			machineSets: []*clusterv1.MachineSet{{
+				Spec: clusterv1.MachineSetSpec{
+					Replicas: &replicas,
+				},
+				Status: clusterv1.MachineSetStatus{
+					Selector:           "",
+					AvailableReplicas:  0,
+					ReadyReplicas:      0,
+					Replicas:           2,
+					ObservedGeneration: 1,
+					ErrorReason:        &msStatusError,
+				},
+			}},
+			newMachineSet: &clusterv1.MachineSet{
+				Spec: clusterv1.MachineSetSpec{
+					Replicas: &replicas,
+				},
+				Status: clusterv1.MachineSetStatus{
+					Selector:           "",
+					AvailableReplicas:  0,
+					ReadyReplicas:      0,
+					Replicas:           2,
+					ObservedGeneration: 1,
+				},
+			},
+			deployment: &clusterv1.MachineDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Generation: 2,
+				},
+				Spec: clusterv1.MachineDeploymentSpec{
+					Replicas: &replicas,
+				},
+			},
+			expectedStatus: clusterv1.MachineDeploymentStatus{
+				ObservedGeneration:  2,
+				Replicas:            2,
+				UpdatedReplicas:     2,
+				ReadyReplicas:       0,
+				AvailableReplicas:   0,
+				UnavailableReplicas: 2,
+				Phase:               "failed",
+			},
+		},
+	}
+
+	for _, e := range statustests {
+		actualStatus := calculateStatus(e.machineSets, e.newMachineSet, e.deployment)
+		if actualStatus != e.expectedStatus {
+			t.Errorf("%s Failed, expected %+v but got %+v", e.desc, e.expectedStatus, actualStatus)
```

Comment:
In case of a test case failure, the message being printed out will be `machine set failed Failed`. Maybe we can print something like `Test Case "machine set failed" Failed`.

Reply:
Please use a go subtest instead and then you don't need to print e.desc

Reply:
changed to subtests